hypothesis-python/src/hypothesis/internal/conjecture/shrinking/floats.py

```diff
@@ -65,18 +65,13 @@ def run_step(self):
         # change that would require shifting the exponent while not changing
         # the float value much.
 
-        for g in [math.floor(self.current), math.ceil(self.current)]:
-            self.consider(g)
-
         if self.consider(int(self.current)):
             self.debug("Just an integer now")
             self.delegate(Integer, convert_to=int, convert_from=float)
             return
 
-        m, n = self.current.as_integer_ratio()
-        i, r = divmod(m, n)
+        # Try dropping precision bits
 
-        # Now try to minimize the top part of the fraction as an integer. This
-        # basically splits the float as k + x with 0 <= x < 1 and minimizes
-        # k as an integer, but without the precision issues that would have.
-        self.call_shrinker(Integer, i, lambda k: self.consider((i * n + r) / n))
+        for p in range(5):
+            for truncate in [math.floor, math.ceil]:
+                self.consider(truncate(self.current * 2**p) / 2**p)
```

Comment:
The crux.

Reply:
Why `5`?  My reading is that `p=0` is a noop, then if `p=1` succeeds we'll be trying higher numbers with a new value of `self.current`.  Should we try reverse-order, no loop at all, or [one of the neat/weird adaptive loops](https://github.com/HypothesisWorks/hypothesis/blob/master/guides/internals.rst#weird-loops)?

Reply:
`p=0` is only noop with regards to shifting, i.e. it rounds up/down to nearest integer.

As for the value 5, that's pragmatic (and also expected to be a discussion point). The highest shift (4) leaves up to 4 decimal digits, although the last two are severely restricted. At some point we should probably just decide "nah, the number is not simple" and give up - trying all up-to-52 shifts is just too much for performance?

Now, the loop: I started out with a variation of the current one, then decided to get fancy by looking for the least-significant set bit and toggle-and-progagate that, then decided that no maintainer would this opaque code for minimal gain. So back to this.

1. We want to try the biggest changes first, to short circuit the loop.
2. Yeah, it doesn't really short circuit as-is; see more-efficient proposal below.
3. I don't think we need adaptive or anything fancy, but if you have neat ideas of how to use it I'm all eyes!

So, if we want to short-circuit properly then maybe something like
```
        for p in range(5):
            floor_val = math.floor(self.current * 2**p) / 2**p
            ceil_val = math.ceil(self.current * 2**p) / 2**p
            if (
                self.consider(floor_val) | self.consider(ceil_val)
            ) or floor_val == ceil_val:
                break
```

Reply:
This would be simplified if we didn't do `ceil`, and I suspect it would be ok since it's shrunk-as-integer next, but I wasn't confident enough to remove it. (The `ceil` survives from the original`)

Reply:
Final comment regarding my item (2).

The reason it didn't short-circuit at this level is that I expected fail-because-not-better to be quick enough to not bother, i.e. it never needs to call down to the predicate. If this is true, maybe the simple version is enough.

Reply:
Ooooh, right, `p=0` is _not_ redundant because we just removed the "try rounding to integer" logic :sweat_smile:

Otherwise, I'd probably try ten instead of five, and leave some comments to the effect that this is an arbitary number chosen as a compromise between completeness and performance; and that we're relying on `self.consider()` skipping test invocation for redundant attempted-shrinks.

Finally, I'd suggest we try round-towards-zero and then away-from-zero rather than ceil/floor; unfortunately [`math.nextafter()`](https://docs.python.org/3/library/math.html#math.nextafter) is new in 3.9 (and we support 3.8) but maybe there's a convenient way to do this, perhaps already in one of our utils modules or compat.py?  If not let's just stick with the status quo.

Reply:
> Finally, I'd suggest we try round-towards-zero and then away-from-zero rather than ceil/floor; unfortunately [`math.nextafter()`](https://docs.python.org/3/library/math.html#math.nextafter) is new in 3.9 (and we support 3.8) but maybe there's a convenient way to do this, perhaps already in one of our utils modules or compat.py? If not let's just stick with the status quo.

I didn't find anything, but `floor` is always towards zero here, since the sign is handled separately. Even so, working on a untouched copy of `self.current` is probably wise, certainly easier to reason about.

Reply:
Ah, of course 🤦‍♂️